httpx/_client.py

```diff
@@ -765,6 +794,10 @@ def _send_handling_auth(
 
         auth_flow = auth.auth_flow(request)
         request = next(auth_flow)
+
+        for hook in self._event_hooks["request"]:
+            hook(request)
```

Comment:
Getting my thoughts down here — are we sure this is the correct/best place to call `request` hooks? AFAICS Requests does not provide a `request` hook so we're in green field here.

There are a few options that I can see…

1. In `.send()` or at the beginning of `_send_handling_redirects()` - Hide redirect requests, auth requests, retries requests.
1. At the beginning of `._send_handling_auth()` - Hide auth requests, retries requests.
1. At the beginning of `._send_single_request()` - Hide retries requests.

As I understand it, currently this PR suggests option 2/.

But interestingly the `response` hook is located at level 1/. So in case of a redirect, we'd have `request` hooks called twice or more, but `response` hooks called once.

Is this appropriate, or do we want `request` being called only once (for the initial request) (which corresponds to option 1/)?

Another data point from the monitoring use case perspective — I can see that the Datadog/Requests tracer integration [wraps `requests.Session.send()`](https://github.com/DataDog/dd-trace-py/blob/master/ddtrace/contrib/requests/connection.py), which _I think_ corresponds to case 1/ above.

Perhaps there's a case to be made about having `request`, `redirect_request`, `auth_request`, and `single_request` hooks, or perhaps just `request` and `single_request`, or…? And then, should we match them up with corresponding `response_*` hooks?

I guess for simplicity we'd really want option 1/, i.e. call `request` hooks before calling `._send_handling_redirects()`, and call `response` hooks on the received resolved response?

Reply:
Resolved via #1267